packages/iframe-coordinator/src/specs/FrameRouter.spec.ts

```diff
@@ -0,0 +1,70 @@
+import { FrameRouterElement } from "../host";
+
+const ENV_DATA = {
+  locale: "en-us",
+  hostRootUrl: "https://example.com/root/",
+};
+
+const ENV_DATA_WITH_HASH = {
+  ...ENV_DATA,
+  hostRootUrl: "https://example.com/root/#/",
+};
+
+describe("The frame router element", () => {
+  beforeAll(() => {
+    window.customElements.define("frame-router", FrameRouterElement);
+  });
+
+  beforeEach(() => {
+    // Reset window hash state
+    window.location.hash = "";
+  });
+
+  /* NOTE: This behavior is inconsistent in practice and needs to be reworked.
+   * For now, these tests prevent regressions against the current inconsistent
+   * behavior.
+   */
+  describe("Host URL management", () => {
+    it("Removes a trailing slash on the host URL if present", () => {
+      const router = new FrameRouterElement();
+      router.clientConfig = {
+        clients: {},
+        envData: ENV_DATA,
+      };
+      //@ts-ignore
+      expect(router._envData).toEqual({
+        ...ENV_DATA,
+        hostRootUrl: "https://example.com/root",
+      });
+    });
+
+    it("Adds a fragment to the tracked host URL if one is present on window but not on the provided root URL", () => {
+      window.location.hash = "foo";
+      const router = new FrameRouterElement();
+      router.clientConfig = {
+        clients: {},
+        envData: ENV_DATA,
+      };
+      //@ts-ignore
+      expect(router._envData).toEqual({
```

Comment:
"_envData" is accessible here? I thought it was a private variable

Reply:
In TypeScript private is only meaningful within the type system. Because TS has to compile to vanilla JS, which has no direct concept of `private`, it just becomes a regular property on the object.  That means it's accessible in tests by disabling the type checks for that line. 

It's not a pattern I'd generally recommend though, and on the next major release when we're doing significant refactors, we should look for ways to make this testable without having to ignore type checks.